synapse/storage/databases/state/bg_updates.py

```diff
@@ -198,7 +217,9 @@ def __init__(self, database: DatabasePool, db_conn, hs):
             columns=["room_id"],
         )
 
-    async def _background_deduplicate_state(self, progress, batch_size):
+    async def _background_deduplicate_state(
+        self, progress: dict, batch_size: int
```

Comment:
I'm not sure if `progress` can be more specific? Maybe `JsonDict`?

Reply:
Not easily. Me and David looked at a similar case. (Though actually it seems reasonable to believe it might be doable as long as you update all background processes at the same time.)

Reply:
I hoped I could define my own `TypedDict`, but this ended up clashing with the background process stuff. Specifically:

- `register_background_update_handler`'s `update_handler` has type `Callable[[JsonDict, int], Awaitable[int]]`
- this means that the `progress` argument type ends up being compared against `JsonDict = Dict[str, Any]`
- `TypedDict` is not a subtype of `JsonDict` because you can add new keys or remove existing keys from a `Dict`
   - so passing `t: TypedDict` to a function `process_dict(d: Dict)` might alter `t`'s keys. This would be bad.
- `TypedDict` is a subtype of `Mapping[str, Any]` because the latter is immutable
- More details here: https://www.python.org/dev/peps/pep-0589/#id20

An `update_handler` itself is called here:

```python
        progress = db_to_json(progress_json)

        time_start = self._clock.time_msec()
        items_updated = await update_handler(progress, batch_size)
```
and `db_to_json` returns a `JsonDict`.

We didn't find a nice way of joining this all up that didn't involve a sea of `cast`s.